### PR TITLE
controller,router: JSON Schema / API Docs

### PR DESCRIPTION
**Pull request by jvatic**
This replaces #175.

- [x] Each schema should reference one or more [examples](https://github.com/flynn/flynn/blob/6553a5fce8bc6637267b45a88b8de67cbec879b4/controller/examples/examples.go#L60-L95) using a custom `examples` keyword. The value should be an array of schema ids referencing example schemas.
- [x] Generate documentation by combining schemas with examples
- [x] Hook up rake task
- [x] Order schemas
- [x] Add expand / collapse toggle for examples
- [x] Add link to docs navigation
- [x] Link to schema JSON

### CODE REVIEW

File: test/test_controller.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	Headers map[string]string `json:"headers,omitempty"`
+	Body    interface{}       `json:"body,omitempty"`
+}
+
+type controllerExample struct {
+	Request  controllerExampleRequest  `json:"request,omitempty"`
+	Response controllerExampleResponse `json:"response,omitempty"`
+}
+
+var jsonContentTypeMatcher = regexp.MustCompile(`\bjson`)
```
Comment (titanous) [inline test/test_controller.go]: s/Matcher/Pattern/

File: test/test_controller.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		ArtifactID: artifact.ID,
+		Env:        map[string]string{"CONTROLLER_KEY": s.clusterConf(t).Key},
+	}
+	t.Assert(client.CreateRelease(release), c.IsNil)
+
+	req := &ct.NewJob{
+		Cmd:       []string{"/bin/flynn-controller-examples"},
+		ReleaseID: release.ID,
+	}
+	rwc, err := client.RunJobAttached(app.ID, req)
```
Comment (titanous) [inline test/test_controller.go]: Use `pkg/exec` to do this, there are examples in other tests.

File: test/test_controller.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	return examples
+}
+
+func (s *ControllerSuite) TestExampleOutput(t *c.C) {
+	examples := s.generateControllerExamples(t)
+	exampleKeys := make([]string, len(examples))
+	i := 0
+	for key, _ := range examples {
+		exampleKeys[i] = key
+		i++
```
Comment (titanous) [inline test/test_controller.go]: Use `append`, and do `make([]string, 0, len(examples))`

File: test/test_controller.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	sort.Strings(exampleKeys)
+	for _, key := range exampleKeys {
+		cacheKey := "https://flynn.io/schema/examples/controller/" + key
+		schema := s.schemaCache[cacheKey]
+		t.Assert(schema, c.Not(c.IsNil))
```
Comment (titanous) [inline test/test_controller.go]: `c.NotNil`

File: website/schema/common.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "https://flynn.io/schema/common#",
+  "definitions": {
+    "id": {
+      "description": "unique identifier",
+      "format": "uuid",
+      "type": "string",
+      "mutable": false
```
Comment (titanous) [inline website/schema/common.json]: Remove these fields for now.

**Comment by titanous** [comment]
@lmars @archSeer This could use another look.

File: script/generate-router-examples [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+usage() {
+  cat <<USAGE >&2
+usage: $0 [options]
+
+OPTIONS:
+  -h            Show this message
+  -i IP         The external IP address [default: the IP assigned to `eth0`]
+  -p PORT       The port to start the resource server on [default: 12345]
+USAGE
```
Comment (titanous) [inline script/generate-router-examples]: I'd rather see a single script that can run any of the example generators instead, of a copied one for each.

File: script/generate-controller-examples [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  ip="${ip:-$(/sbin/ifconfig eth0 \
+    | grep -oP 'inet addr:\S+' \
+    | cut -d: -f2)}"
+  port="${port:-"12345"}"
+
+  export CONTROLLER_KEY="s3cr3t"
+
+  info "bootstrapping flynn" >&2
+  cluster_add=$("${ROOT}/script/bootstrap-flynn" &> >(tee /dev/stderr) | grep "flynn cluster add" | tail -1)
```
Comment (lmars) [inline script/generate-controller-examples]: This needs the same fix as 3638155 (and maybe pull it out into a shared function)

File: test/test_controller.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	var schemaPaths []string
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() {
+			schemaPaths = append(schemaPaths, path)
+		}
+		return nil
+	}
+	schemaRoot, err := filepath.Abs(filepath.Join("..", "website", "schema"))
+	t.Assert(err, c.IsNil)
+	filepath.Walk(schemaRoot, walkFn)
```
Comment (lmars) [inline test/test_controller.go]: The returned error should be checked

**Comment by jvatic** [comment]
@titanous @lmars comments addressed

File: website/schema/common.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      "format": "date-time",
+      "type": "string"
+    },
+    "updated_at": {
+      "description": "object last updated timestamp",
+      "format": "date-time",
+      "type": "string"
+    },
+    "config": {
+      "type": "object"
```
Comment (jvatic) [inline website/schema/common.json]: Specify what the properties may look like

File: website/schema/controller/common.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    "cmd": {
+      "description": "shell command",
+      "type": "array",
+      "items": {
+        "type": "string"
+      }
+    },
+    "env": {
+      "description": "environment variables",
+      "type": "object"
```
Comment (jvatic) [inline website/schema/controller/common.json]: Ditto

File: script/generate-examples [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      ?)
+        usage
+        exit 2
+        ;;
+    esac
+  done
+
+  ip="${ip:-$(/sbin/ifconfig eth0 \
+    | grep -oP 'inet addr:\S+' \
+    | cut -d: -f2)}"
```
Comment (titanous) [inline script/generate-examples]: This should not be necessary: https://github.com/flynn/flynn/commit/eb3ad9be50e3f42c69a0e6e7936e47a231d50a02

File: script/generate-examples [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+  info "generating ${component} examples..." >&2
+  docker run \
+    -e "CONTROLLER_KEY=${CONTROLLER_KEY}" \
+    -e "DISCOVERD=${ip}:1111" \
+    -e "PORT=${port}" \
+    -p "${port}:${port}" \
+    "flynn/${component}-examples" \
+    "/bin/flynn-${component}-examples" \
+    2>&1
```
Comment (titanous) [inline script/generate-examples]: Use `flynn-host run` here.

File: website/source/layouts/docs.erb [hunk 13]
```diff
@@ -17,6 +17,8 @@
             <%= nav_link_with_active("Contributing", "/docs/contributing") %>
             <%= nav_link_with_active("Development", "/docs/development") %>
             <%= nav_link_with_active("Command Line", "/docs/cli") %>
+            <%= nav_link_with_active("Controller API", "/docs/controller") %>
+            <%= nav_link_with_active("Router API", "/docs/router") %>
```
Comment (titanous) [inline website/source/layouts/docs.erb]: Maybe `/docs/api/router`?

File: script/generate-examples [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+  if [[ "${cluster_add:0:17}" != "flynn cluster add" ]]; then
+    fail "Bootstrap failed"
+  fi
+
+  info "generating ${component} examples..." >&2
+
+  imageid=`sudo docker images --no-trunc -q flynn/${component}-examples | tr -d '\r\n'`
+  imageurl="https://registry.hub.docker.com/flynn/${component}-examples?id=$imageid"
+
+  host/bin/flynn-host run $imageurl sh -c "CONTROLLER_KEY=${CONTROLLER_KEY} PORT=${port} /bin/flynn-${component}-examples" 2>&1
```
Comment (titanous) [inline script/generate-examples]: `${ROOT}/bin/flynn-host`

File: script/generate-examples [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+ROOT="$(cd "$(dirname "$0")/.." && pwd)"
+source "${ROOT}/script/lib/ui.sh"
+
+usage() {
+  cat <<USAGE >&2
+usage: $0 <controller|router> [options]
+
+OPTIONS:
+  -h            Show this message
+  -i IP         The external IP address [default: the IP assigned to eth0]
```
Comment (titanous) [inline script/generate-examples]: This argument isn't used.

**Comment by titanous** [comment]
Please add an example for the deploy endpoint, and make sure you update the examples for both.

**Comment by jvatic** [comment]
@titanous comments addressed

File: Godeps/Godeps.json [hunk 16]
```diff
@@ -111,11 +110,6 @@
 			"Rev": "034ba8b88fdcc3d5100ae0d2edc914673b15a2f4"
 		},
 		{
-			"ImportPath": "github.com/docker/docker/pkg/parsers",
-			"Comment": "v1.4.1-165-g034ba8b",
-			"Rev": "034ba8b88fdcc3d5100ae0d2edc914673b15a2f4"
-		},
-		{
```
Comment (titanous) [inline Godeps/Godeps.json]: These packages shouldn't be deleted, manually undo, or do the godep save from Linux.

File: pkg/httprecorder/httprecorder.go [hunk 17]
```diff
@@ -28,7 +28,11 @@ type CompiledRequest struct {
 
 func NewWithClient(client *http.Client) *Recorder {
 	r := &Recorder{Client: client}
-	r.originalTransport = client.Transport
+	if client.Transport != nil {
+		r.originalTransport = client.Transport
+	} else {
+		r.originalTransport = &http.Transport{}
+	}
```
Comment (titanous) [inline pkg/httprecorder/httprecorder.go]:
```go
r.originalTransport = client.Transport
if r.originalTransport == nil {
...
```

**Comment by titanous** [comment]
LGTM

**Comment by titanous** [comment]
:balloon:

**Pull request merged**